src/main/java/com/avispl/symphony/dal/infrastructure/management/qsc/qsysreflect/QSysReflectCommunicator.java

```diff
@@ -0,0 +1,734 @@
+	 */
+	private void retrieveDevices() {
+		try {
+			Map<String, PropertiesMapping> mapping = new PropertiesMappingParser().loadYML("qsysreflect/model-mapping.yml", getClass());
+			AggregatedDeviceProcessor aggregatedDeviceProcessor = new AggregatedDeviceProcessor(mapping);
+			String responseDeviceList = this.doGet(QSysReflectConstant.QSYS_URL_CORES, String.class);
+			JsonNode devices = objectMapper.readTree(responseDeviceList);
+			for (int i = 0; i < devices.size(); i++) {
+				JsonNode currentDevice = devices.get(i);
+				deviceStatusMessageMap.put(currentDevice.get(QSysReflectConstant.ID).asText(), currentDevice.get(QSysReflectConstant.STATUS)
+						.get(QSysReflectConstant.MESSAGE).asText());
+			}
+			aggregatedDeviceList = new ArrayList<>(aggregatedDeviceProcessor.extractDevices(devices));
+		} catch (Exception e) {
+			String errorMessage = "Aggregated Device Data Retrieval-Error:"
+					+ e.getMessage() + " with cause: " + e.getCause().getMessage();
+			deviceErrorMessagesList.add(errorMessage);
+			logger.error(errorMessage);
+		}
+	}
+
+	/**
+	 * Get system information every 30 seconds
+	 * API Endpoint: /systems
+	 * Success: return list of systems within the organization
+	 */
+	public void retrieveSystemInfo() {
+		// Retrieve system information every 30 seconds
+		try {
+			String systemResponse = this.doGet(QSysReflectConstant.QSYS_URL_SYSTEMS, String.class);
+			JsonNode systems = objectMapper.readTree(systemResponse);
+			systemResponseList.clear();
+			for (int i = 0; i < systems.size(); i++) {
+				SystemResponse sysRes = objectMapper.treeToValue(systems.get(i), SystemResponse.class);
+				systemResponseList.add(sysRes);
+			}
+		} catch (Exception e) {
+			String errorMessage = "System Information Data Retrieval-Error:"
+					+ e.getMessage() + " with cause: " + e.getCause().getMessage();
+			systemErrorMessagesList.add(errorMessage);
+			logger.error(errorMessage);
+		}
+	}
+
+	/**
+	 * Populate data to statistics
+	 *
+	 * @param stats Map of statistic
+	 */
+	private void populateSystemData(Map<String, String> stats) {
+		if (!systemResponseList.isEmpty()) {
+			if (logger.isDebugEnabled()) {
+				logger.debug("Populating system information data");
+			}
+			List<AggregatedDevice> resultAggregatedDeviceList = getFilteredAggregatedDeviceList();
+			Map<String, String> deviceNameAndModelMap = new HashMap<>();
+			for (AggregatedDevice aggregatedDevice : resultAggregatedDeviceList) {
+				deviceNameAndModelMap.put(aggregatedDevice.getDeviceName(), aggregatedDevice.getDeviceModel());
+			}
+			synchronized (systemResponseList) {
+				for (SystemResponse systemResponse : systemResponseList) {
+					// Apply filter for system information that match the devices' name
+					if (deviceNameAndModelMap.containsKey(systemResponse.getCoreName())) {
+						stats.put(systemResponse.getName() + QSysReflectConstant.HASH + QSysReflectSystemMetric.SYSTEM_ID.getName(), String.valueOf(systemResponse.getId()));
```

Comment:
Instead of concatenating a + b + c you might have used string template of "%s#%s" here for string.format()

Not something critical, but this would help you avoiding adding a single characters into constants + would make it a bit less human-error prone and a bit more readable :)

Reply:
Updated